neutrons/SNAPRed
# Ewm8373 rebin normalization and reduction output correctly

## Description of work

This pr adds appropriate binning for both the artificial normalization subflow of Reduction as well as the Reduction outputs proper.

This is important because the Divide operation along with several others are sensitive to the current binning of an EventWorkspaces, even before we drop events and make it a Histogram(solidifying the resolution/bins).

It also sets the outputs of reduction to distributions, as that is what a user will expect of their output data.

## Explanation of work

- I refactored the way we bin group focussed data into a standardized Recipe, `RebinFocussedGroupDataRecipe`
- Moved the Artificial Normalization code into `GenerateFocussedVanadiumRecipe` as that was originally where we wanted to put it.  It helps keeping like with like.
- `ArtificialNormalizationRequest` should not be arbitrarily generating the output workspace name
- `ApplyNormalizationRecipe` now uses the centralized rebinning, as well as rebinning before a Divide and dropping events after the Divide
- Added `RebinFocussedGroupDataRecipe` to house binning for Reduction/Normalization rebin operations.
- Updated `Recipe` to be able to `adopt` host algorithm (Still needs further investigation for proper use as it seems to not share a mantidsnapper instance, but should be fine to include)
- `ReductionGroupProcessingRecipe` now respects the outputworkspace property
- `ReductionRecipe` updated to account for new way of generating an Artificial Normalization
- `CreateArtificialNormalizationAlgo` input properties refactored to use Mantid primatives instead of ingredients (i.e. `int`),  only clone if outputworkspace is different than input, convert units before we drop events.
- `ReductionService` changed to rebin as part of the artificial norm preview
- `WorkspaceNameGenerator` added new name for art norm preview
 
## To test

This is heavily dependent on your ability to diagnose the data and confirm binning.
So we will rely on Malcolm to confirm whether or not this data output is correct.

As a dev, do the best you can to confirm that the new operation completed successfully.

Freshly init state folder for desired run, I have been using `59039`
Perform reduction on `59039`

### On `Next`

Plot the artificial norm workspace, observe the results.
Reduce your data.
Plot the output `column` workspace.
Observe the data.
Observe that it is not a distribution (i.e. the min y value is 0 and not 1)
Observe any other artifact.

### On `This Branch`

Plot the Artificial Norm workspace, observe the different results.
![image](https://github.com/user-attachments/assets/bc9e62b8-e182-4c60-bdad-b755d76ded7e)
If you double the smoothing and peak window values you should get something much smoother
![image](https://github.com/user-attachments/assets/f09e1a4d-83b8-4c10-9722-6423ab5416af)

Observe the Logs, confirm that the log "Rebinning workspace for group..." now appears where on next it doesnt.
Reduce your data.
Plot the output `column` workspace.
Observe the data.
Observe that it is a distribution (i.e. the min y value is 1)
Observe a relatively flat output.
![image](https://github.com/user-attachments/assets/9ec94b47-a541-49dc-978d-a115cb524a3d)

## Link to EWM item


[EWM#8373](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=8373)

### Verification




### Acceptance Criteria



This list is for ease of reference, and does not replace reading the EWM story as part of the review.  Verify this list matches the EWM story before reviewing.

- [x] the input datasets for peakClip must be rebinned following the pixelGroupingParameters dMin, dMax and dBin which SNAPRed determines for each individual pixel group before calculating the normalization function in peakClip.
- [x] The outcome is that the Divide output workspace flag `isDistribution` is set equal to False, when it should be True.
